lib/require_dependency_backward_compatibility.rb

```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# Patch `require_dependency` to maintain backward compatibility with some
+# plugins. Calls to `require_dependency` are deprecated and we should remove
+# them whenever possible.
+module RequireDependencyBackwardCompatibility
+  def require_dependency(filename)
+    name = filename.to_s
+    return if name == 'jobs/base'
+    return super(name.sub(/^lib\//, '')) if name.start_with?('lib/')
+    super
+  end
```

Comment:
Would it be possible to expand on the problems with the `jobs` directory such that we have to carry this?

Reply:
Yes, it has to do with the “new” autoload mechanism and how we originally organised jobs. We have our jobs under `app/jobs` so that’s where you expect them in a typical Rails app but what you don’t expect is for them to start with a `Jobs` namespace. The Rails Way™ for stuff under `app` is to add a suffix based on the directory where files are. So to continue with jobs, we should have a `GrantEmojiJob` defined in `app/jobs/grant_emoji_job.rb` and not `Jobs::GrantEmoji` defined in `app/jobs/onceoff/grant_emoji.rb`. On top of that we have a `jobs/base.rb` file that defines a module named `Jobs` and a class named `Jobs::Base` whereas it should just define `Base` (based on its filename). But to follow Rails conventions we should probably have an `ApplicationJob` class and all other jobs should inherit from it. This would be defined in a file named `app/jobs/application_job.rb`.

We could (and probably should at some point) rename/move around our jobs but by doing this in a naive way (just rename all the things) we’d have some problems: enqueued jobs won’t be processed as their names would change and it seems a lot of plugins are requiring `jobs/base` so it will break them too.

`require` and `require_dependency` are already managed/patched by Zeitwerk so we have to play nice with it from now on. As our jobs are autoloaded anyway, we don’t really need to require them anywhere (even in plugins) so with this small patch when we encounter `jobs/base` we simply ignore it, relying on Zeitwerk to do its work properly (otherwise it will crash). More or less the same applies for requires starting by `lib/` (which was clearly a bug at some point), we still have some code using that so we simply remove the `lib/` prefix.

I hope it’s a bit clearer, if not, just tell me :slightly_smiling_face:

Reply:
Thank you for the detailed explanation here :+1: Yea I agree now is probably not the best time to mess with the directory structure for jobs. Let's live with this patch now but I think it'll be super helpful if we can improve the comments in this file with the problems you described above. After that, we can then bring this discussion to dev to see if we should invest time fixing our job files to properly follow the autoloading convenetions.